src/liballoc/collections/btree/node.rs

```diff
@@ -158,29 +158,32 @@ impl<K, V> InternalNode<K, V> {
 }
 
 /// An owned pointer to a node. This basically is either `Box<LeafNode<K, V>>` or
-/// `Box<InternalNode<K, V>>`. However, it contains no information as to which of the two types
-/// of nodes is actually behind the box, and, partially due to this lack of information, has no
-/// destructor.
+/// `Box<InternalNode<K, V>>` or `Box<NodeHeader<(), ())` (i.e. EMPTY_ROOT_NODE).
```

Comment:
The shared root is neither boxed nor uniquely owned by one `BoxedNode`, so although it's good to note that this pointer can refer to the shared root, wording it like this is misleading IMO. I'd suggest the following outline instead:

- This is usually an owned pointer to a LeafNode/InternalNode, but it can also be a pointer (no ownership) to EMPTY_ROOT_NODE
- No destructor in part because we don't know which of these variants it is
- All of these types have a `NodeHeader<K, V>` prefix (even `EMPTY_ROOT: NodeHeader<(), ()>`), so that's the pointee type we store

Reply:
I'm confused by and can't find a description of "prefix" in a pointer context. Does that mean the data it points at always has at least the size and layout of `NodeHeader<K, V>`? Not that I know any better word for it. Does "prefix" also imply that pointer alignment is at least as high as that of `NodeHeader<K, V>`?

Reply:
To be precise, I mean that all of `InternalNode<K, V>`, `LeafNode<K, V>` and `NodeHeader<(), ()>`:

- are at least as large as `NodeHeader<K, V>`
- have alignment at least as large as `NodeHeader<K, V>`
- store the same kinds of data, at the same offsets, as a `NodeHeader<K, V>` in the first `size_of::<NodeHeader<K, V>>`

This is the reason why `as_header()` is unconditionally safe. Alignment in particular is probably worth calling out separately, as it's not really implied by "prefix" (e.g., you might say `i32` is a prefix of `#[repr(C)] struct { a: i32, b: i64 }` but alignment differs).